keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -215,6 +225,113 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
+    return deliveryConfig.environments.map { environment ->
+      val artifactVersions = deliveryConfig.artifacts.map { artifact ->
+        //          select e.name, eav.artifact_version, eav.approved_at, eav.deployed_at
+//          from environment_artifact_versions eav, delivery_artifact da, environment e, delivery_config dc
+//          where da.uid = eav.artifact_uid
+//          and da.name = 'keel'
+//          and da.type = 'DEB'
+//          and e.uid = eav.environment_uid
+//          and e.delivery_config_uid = dc.uid
+//          and dc.name = 'keel-manifest'
+//          union
+//          select e.name, dav.version, null, null
+//          from delivery_artifact_version dav, delivery_artifact da, environment e, delivery_config dc
+//          where da.uid = dav.delivery_artifact_uid
+//          and da.name = 'keel'
+//          and da.type = 'DEB'
+//          and e.delivery_config_uid = dc.uid
+//          and dc.name = 'keel-manifest'
+//          and not exists (
+//          select 1 from environment_artifact_versions eav
+//          where eav.artifact_version = dav.version
+//          and eav.environment_uid = e.uid
+//        )
+        val rawResults = jooq
```

Comment:
This query is so large and confusing. Could you put a function description that describes the steps you're taking to build this information?